### PR TITLE
store/tikv: udpate backoff.

### PR DESCRIPTION
**Pull request by disksing**
This PR contains a new approach to do retry. A `Backoffer` struct is created with each kv API call, the struct will track every retryable error occurs within the call and throw an error if total sleep time exceeds limit. 

@shenli @zimulala @coocood

### CODE REVIEW

File: store/tikv/backoff.go [hunk 1]
```diff
@@ -59,12 +57,88 @@ func NewBackoff(retry, base, cap, jitter int) func() error {
...
+	case boPDRPC:
+		return NewBackoffFn(500, 3000, EqualJitter)
+	case boRegionMiss:
+		return NewBackoffFn(100, 500, NoJitter)
+	}
+	return nil
+}
+
+const (
+	copBuildTaskMaxBackoff = 3000
```
Comment (shenli) [inline store/tikv/backoff.go]: This is max retry time?

File: store/tikv/backoff.go [hunk 2]
```diff
@@ -59,12 +57,88 @@ func NewBackoff(retry, base, cap, jitter int) func() error {
...
+// NewBackoff creates a Backoff with maximum sleep time(in ms).
+func NewBackoff(maxSleep int) *Backoff {
+	return &Backoff{
+		maxSleep: maxSleep,
+	}
+}
+
+// Backoff sleeps a while base on the backoffType and records the error message.
+// It returns an retryable error if total sleep time exceeds maxSleep.
+func (b *Backoff) Backoff(typ backoffType, err string) error {
```
Comment (shenli) [inline store/tikv/backoff.go]: Why not use error as the second argument?

File: store/tikv/backoff.go [hunk 3]
```diff
@@ -59,12 +57,88 @@ func NewBackoff(retry, base, cap, jitter int) func() error {
...
+
+// NewBackoff creates a Backoff with maximum sleep time(in ms).
+func NewBackoff(maxSleep int) *Backoff {
+	return &Backoff{
+		maxSleep: maxSleep,
+	}
+}
+
+// Backoff sleeps a while base on the backoffType and records the error message.
+// It returns an retryable error if total sleep time exceeds maxSleep.
```
Comment (zimulala) [inline store/tikv/backoff.go]: an -> a ?

File: store/tikv/backoff.go (hunk 1 above)
Comment (disksing) [inline store/tikv/backoff.go]: It's the maximum backoff sleep time. I'll add a comment here.

File: store/tikv/backoff.go (hunk 2 above)
Comment (disksing) [inline store/tikv/backoff.go]: Maybe error is better. I'll fix it.

File: store/tikv/coprocessor.go [hunk 4]
```diff
@@ -207,6 +208,7 @@ type copIterator struct {
 
 // Pick the next new copTask and send request to tikv-server.
 func (it *copIterator) work() {
+	bo := NewBackoff(copNextMaxBackoff)
```
Comment (shenli) [inline store/tikv/coprocessor.go]: Add a TODO here. We should reconsider this.

File: store/tikv/kv.go [hunk 5]
```diff
@@ -124,46 +124,45 @@ func (s *tikvStore) UUID() string {
 }
 
 func (s *tikvStore) CurrentVersion() (kv.Version, error) {
-	startTS, err := s.getTimestampWithRetry()
+	bo := NewBackoff(storeVersionMaxBackoff)
```
Comment (shenli) [inline store/tikv/kv.go]: We can put the backoff into getTimestampWithRetry().

Comment (shenli) [inline store/tikv/kv.go]: Ignore it.

File: store/tikv/backoff.go (hunk 2 above)
Comment (coocood) [inline store/tikv/backoff.go]: The function name is the same as the type name, which is confusing.
How about rename the type to `Backoffer`?

Comment (disksing) [inline store/tikv/backoff.go]: @coocood Good idea.

File: store/tikv/region_cache.go [hunk 6]
```diff
@@ -196,17 +196,21 @@ func (c *RegionCache) dropRegionFromCache(verID RegionVerID) {
 }
 
 // loadRegion loads region from pd client, and picks the first peer as leader.
-func (c *RegionCache) loadRegion(key []byte) (*Region, error) {
-	var region *Region
-	var backoffErr error
-	for backoff := pdBackoff(); backoffErr == nil; backoffErr = backoff() {
+func (c *RegionCache) loadRegion(bo *Backoffer, key []byte) (*Region, error) {
```
Comment (zimulala) [inline store/tikv/region_cache.go]: Add an error declaration to reduce the invoking of  `Backoff` and error judgments.
```
if err！= nil {
   if err =  bo.Backoff(boPDRPC, err); err != nil {
       return nil, errors.Trace(err)
   }
}
```

File: store/tikv/txn_committer.go [hunk 7]
```diff
@@ -110,29 +110,29 @@ func (c *txnCommitter) iterKeys(keys [][]byte, f func(batchKeys) error, sizeFn f
...
+		err = c.doBatches(bo, batches[:1], f)
 		if err != nil {
 			return errors.Trace(err)
 		}
 		batches = batches[1:]
 	}
 	if asyncNonPrimary {
 		go func() {
-			c.doBatches(batches, f)
+			c.doBatches(bo, batches, f)
```
Comment (zimulala) [inline store/tikv/txn_committer.go]: bo -> bo.Fork() ?

Comment (disksing) [inline store/tikv/txn_committer.go]: @zimulala Seems we don't have to fork here. The original goroutine will not use `bo` any more, so it should be safe.

**Comment by zimulala** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Comment by disksing** [comment]
PTAL @shenli

**Comment by shenli** [comment]
LGTM

**Pull request merged**